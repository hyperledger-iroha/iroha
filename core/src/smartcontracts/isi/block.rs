//! This module contains trait implementations related to block queries
use eyre::{Result, WrapErr};
use iroha_data_model::{
    evaluate::ExpressionEvaluator,
    query::{
        block::FindBlockHeaderByHash,
        error::{FindError, QueryExecutionFailure},
    },
};
use iroha_telemetry::metrics;

use super::*;

impl ValidQuery for FindAllBlocks {
    #[metrics(+"find_all_blocks")]
    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, QueryExecutionFailure> {
        let blocks = wsv.all_blocks_by_value().rev().collect();
        Ok(blocks)
    }
}

impl ValidQuery for FindAllBlockHeaders {
    #[metrics(+"find_all_block_headers")]
    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, QueryExecutionFailure> {
        let block_headers = wsv
            .all_blocks_by_value()
            .rev()
            .map(|block| block.into_v1().header)
            .collect();
        Ok(block_headers)
    }
}

impl ValidQuery for FindBlockHeaderByHash {
    #[metrics(+"find_block_header")]
    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, QueryExecutionFailure> {
        let hash = wsv
            .evaluate(&self.hash)
            .wrap_err("Failed to evaluate hash")
            .map_err(|e| QueryExecutionFailure::Evaluate(e.to_string()))?
            .typed();

        let block = wsv
            .all_blocks_by_value()
            .find(|block| block.hash() == hash)
            .ok_or_else(|| QueryExecutionFailure::Find(Box::new(FindError::Block(hash))))?;

        Ok(block.into_v1().header)
    }
}
