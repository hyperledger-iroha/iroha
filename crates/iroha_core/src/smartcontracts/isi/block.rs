//! This module contains trait implementations related to block queries
use eyre::Result;
use iroha_data_model::query::{
    error::QueryExecutionFail,
    predicate::{
        predicate_atoms::block::{BlockHeaderPredicateBox, SignedBlockPredicateBox},
        CompoundPredicate,
    },
};
use iroha_telemetry::metrics;
use nonzero_ext::nonzero;

use super::*;
use crate::{smartcontracts::ValidQuery, state::StateReadOnly};

impl ValidQuery for FindBlocks {
    #[metrics(+"find_blocks")]
    fn execute(
        self,
        filter: CompoundPredicate<SignedBlockPredicateBox>,
        state_ro: &impl StateReadOnly,
    ) -> Result<impl Iterator<Item = Self::Item>, QueryExecutionFail> {
        Ok(state_ro
            .all_blocks(nonzero!(1_usize))
            .rev()
            .filter(move |block| filter.applies(block))
            .map(|block| (*block).clone()))
    }
}

impl ValidQuery for FindBlockHeaders {
    #[metrics(+"find_block_headers")]
    fn execute(
        self,
        filter: CompoundPredicate<BlockHeaderPredicateBox>,
        state_ro: &impl StateReadOnly,
    ) -> Result<impl Iterator<Item = Self::Item>, QueryExecutionFail> {
        Ok(state_ro
            .all_blocks(nonzero!(1_usize))
            .rev()
            .filter(move |block| filter.applies(block.header()))
            .map(|block| block.header().clone()))
    }
}
