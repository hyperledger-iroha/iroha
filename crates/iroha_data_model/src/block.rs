//! This module contains `Block` and related implementations.
//!
//! `Block`s are organised into a linear sequence over time (also known as the block chain).

#[cfg(not(feature = "std"))]
use alloc::{
    boxed::Box,
    collections::{BTreeMap, BTreeSet},
    format,
    string::String,
    vec::Vec,
};
use core::{fmt::Display, time::Duration};
#[cfg(feature = "std")]
use std::collections::{BTreeMap, BTreeSet};

use derive_more::Display;
use iroha_crypto::{HashOf, MerkleTree, SignatureOf};
use iroha_data_model_derive::model;
use iroha_macro::FromVariant;
use iroha_schema::IntoSchema;
use iroha_version::{declare_versioned, version_with_scale};
use parity_scale_codec::{Decode, Encode};
use serde::{Deserialize, Serialize};

pub use self::model::*;
use crate::transaction::{error::TransactionRejectionReason, prelude::*};

#[model]
mod model {
    use core::num::NonZeroU64;

    use getset::{CopyGetters, Getters};

    use super::*;

    #[derive(
        Debug,
        Display,
        Clone,
        Copy,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        CopyGetters,
        Getters,
        Decode,
        Encode,
        Deserialize,
        Serialize,
        IntoSchema,
    )]
    #[display(fmt = "{} (â„–{height})", "self.hash()")]
    #[allow(missing_docs)]
    #[ffi_type]
    pub struct BlockHeader {
        /// Number of blocks in the chain including this block.
        #[getset(get_copy = "pub")]
        pub height: NonZeroU64,
        /// Hash of the previous block in the chain.
        #[getset(get_copy = "pub")]
        pub prev_block_hash: Option<HashOf<BlockHeader>>,
        /// Hash of merkle tree root of transactions' hashes.
        /// None if no transactions (empty block).
        #[getset(get_copy = "pub")]
        pub transactions_hash: Option<HashOf<MerkleTree<SignedTransaction>>>,
        /// Creation timestamp (unix time in milliseconds).
        #[getset(skip)]
        pub creation_time_ms: u64,
        /// Value of view change index. Used to resolve soft forks.
        #[getset(skip)]
        pub view_change_index: u32,
    }

    #[derive(
        Debug,
        Display,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Encode,
        Serialize,
        IntoSchema,
        Deserialize,
        Decode,
    )]
    #[display(fmt = "({header})")]
    #[allow(missing_docs)]
    #[allow(clippy::redundant_pub_crate)]
    pub(crate) struct BlockPayload {
        /// Block header
        pub header: BlockHeader,
        /// array of transactions, which successfully passed validation and consensus step.
        pub transactions: Vec<SignedTransaction>,
    }

    /// Signature of a block
    #[derive(
        Debug,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Decode,
        Encode,
        Deserialize,
        Serialize,
        IntoSchema,
    )]
    pub struct BlockSignature(
        /// Index of the peer in the topology
        pub u64,
        /// Payload
        pub SignatureOf<BlockHeader>,
    );

    /// Signed block
    #[version_with_scale(version = 1, versioned_alias = "SignedBlock")]
    #[derive(
        Debug,
        Display,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Encode,
        Serialize,
        IntoSchema,
        Decode,
        Deserialize,
    )]
    #[display(fmt = "{}", "self.header()")]
    #[ffi_type]
    pub struct SignedBlockV1 {
        /// Signatures of peers which approved this block.
        pub(super) signatures: BTreeSet<BlockSignature>,
        /// Block payload
        pub(super) payload: BlockPayload,
        /// Collection of rejection reasons for every transaction if exists
        ///
        /// # Warning
        ///
        /// Transaction errors are not part of the block hash or protected by the block signature.
        pub(super) errors: BTreeMap<u64, TransactionRejectionReason>,
    }
}

#[cfg(any(feature = "ffi_export", feature = "ffi_import"))]
declare_versioned!(SignedBlock 1..2, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, FromVariant, iroha_ffi::FfiType, IntoSchema);
#[cfg(all(not(feature = "ffi_export"), not(feature = "ffi_import")))]
declare_versioned!(SignedBlock 1..2, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, FromVariant, IntoSchema);

impl BlockHeader {
    /// Checks if it's a header of a genesis block.
    #[inline]
    pub const fn is_genesis(&self) -> bool {
        self.height.get() == 1
    }

    /// Creation timestamp
    pub const fn creation_time(&self) -> Duration {
        Duration::from_millis(self.creation_time_ms)
    }

    /// Calculate block hash
    #[inline]
    pub fn hash(&self) -> HashOf<BlockHeader> {
        iroha_crypto::HashOf::new(self)
    }
}

impl SignedBlockV1 {
    fn hash(&self) -> HashOf<BlockHeader> {
        self.payload.header.hash()
    }

    fn header(&self) -> BlockHeader {
        self.payload.header
    }
}

impl SignedBlock {
    /// Create new block with a given signature
    ///
    /// # Warning
    ///
    /// All transactions are categorized as valid
    #[cfg(feature = "transparent_api")]
    pub fn presigned(
        signature: BlockSignature,
        header: BlockHeader,
        transactions: Vec<SignedTransaction>,
    ) -> SignedBlock {
        SignedBlockV1 {
            signatures: [signature].into_iter().collect(),
            payload: BlockPayload {
                header,
                transactions,
            },
            errors: BTreeMap::new(),
        }
        .into()
    }

    /// Setter for transaction errors
    #[cfg(feature = "transparent_api")]
    pub fn set_transaction_errors(
        &mut self,
        errors: BTreeMap<u64, TransactionRejectionReason>,
    ) -> &mut Self {
        let SignedBlock::V1(block) = self;
        block.errors = errors;
        self
    }

    /// Return error for the transaction index
    pub fn error(&self, tx: usize) -> Option<&TransactionRejectionReason> {
        let SignedBlock::V1(block) = self;
        block.errors.get(&(tx as u64))
    }

    /// Block payload. Used for tests
    #[cfg(feature = "transparent_api")]
    pub fn payload(&self) -> &BlockPayload {
        let SignedBlock::V1(block) = self;
        &block.payload
    }

    /// Block header
    #[inline]
    pub fn header(&self) -> BlockHeader {
        let SignedBlock::V1(block) = self;
        block.header()
    }

    /// Signatures of peers which approved this block.
    #[inline]
    pub fn signatures(
        &self,
    ) -> impl ExactSizeIterator<Item = &BlockSignature> + DoubleEndedIterator {
        let SignedBlock::V1(block) = self;
        block.signatures.iter()
    }

    /// Block transactions
    #[inline]
    pub fn transactions(&self) -> impl ExactSizeIterator<Item = &SignedTransaction> {
        let SignedBlock::V1(block) = self;
        block.payload.transactions.iter()
    }

    /// Check if block is empty (has no transactions)
    #[inline]
    pub fn is_empty(&self) -> bool {
        let SignedBlock::V1(block) = self;
        block.payload.transactions.is_empty()
    }

    /// Collection of rejection reasons for every transaction if exists
    ///
    /// # Warning
    ///
    /// Transaction errors are not part of the block hash or protected by the block signature.
    pub fn errors(&self) -> impl ExactSizeIterator<Item = (&u64, &TransactionRejectionReason)> {
        let SignedBlock::V1(block) = self;
        block.errors.iter()
    }

    /// Calculate block hash
    #[inline]
    pub fn hash(&self) -> HashOf<BlockHeader> {
        let SignedBlock::V1(block) = self;
        block.hash()
    }

    /// Add additional signature to this block
    #[cfg(feature = "transparent_api")]
    pub fn sign(&mut self, private_key: &iroha_crypto::PrivateKey, signatory: usize) {
        let SignedBlock::V1(block) = self;

        block.signatures.insert(BlockSignature(
            signatory as u64,
            SignatureOf::new(private_key, &block.payload.header),
        ));
    }

    /// Add signature to the block
    ///
    /// # Errors
    ///
    /// if signature is invalid
    #[cfg(feature = "transparent_api")]
    pub fn add_signature(&mut self, signature: BlockSignature) -> Result<(), iroha_crypto::Error> {
        if self.signatures().any(|s| signature.0 == s.0) {
            return Err(iroha_crypto::Error::Signing(
                "Duplicate signature".to_owned(),
            ));
        }

        let SignedBlock::V1(block) = self;
        block.signatures.insert(signature);

        Ok(())
    }

    /// Replace signatures without verification
    ///
    /// # Errors
    ///
    /// if there is a duplicate signature
    #[cfg(feature = "transparent_api")]
    pub fn replace_signatures(
        &mut self,
        signatures: BTreeSet<BlockSignature>,
    ) -> Result<BTreeSet<BlockSignature>, iroha_crypto::Error> {
        if signatures.is_empty() {
            return Err(iroha_crypto::Error::Signing("Signatures empty".to_owned()));
        }

        signatures.iter().map(|signature| signature.0).try_fold(
            BTreeSet::new(),
            |mut acc, elem| {
                if !acc.insert(elem) {
                    return Err(iroha_crypto::Error::Signing(format!(
                        "{elem}: Duplicate signature"
                    )));
                }

                Ok(acc)
            },
        )?;

        let SignedBlock::V1(block) = self;
        Ok(core::mem::replace(&mut block.signatures, signatures))
    }

    /// Creates genesis block signed with genesis private key (and not signed by any peer)
    #[cfg(feature = "std")]
    pub fn genesis(
        transactions: Vec<SignedTransaction>,
        private_key: &iroha_crypto::PrivateKey,
    ) -> SignedBlock {
        use nonzero_ext::nonzero;

        let transactions_hash = transactions
            .iter()
            .map(SignedTransaction::hash)
            .collect::<MerkleTree<_>>()
            .root()
            .expect("Genesis block must have transactions");
        let creation_time_ms = Self::get_genesis_block_creation_time(&transactions);
        let header = BlockHeader {
            height: nonzero!(1_u64),
            prev_block_hash: None,
            transactions_hash: Some(transactions_hash),
            creation_time_ms,
            view_change_index: 0,
        };

        let signature = BlockSignature(0, SignatureOf::new(private_key, &header));
        let payload = BlockPayload {
            header,
            transactions,
        };

        SignedBlockV1 {
            signatures: [signature].into_iter().collect(),
            payload,
            errors: BTreeMap::new(),
        }
        .into()
    }

    #[cfg(feature = "std")]
    fn get_genesis_block_creation_time(transactions: &[SignedTransaction]) -> u64 {
        use std::time::SystemTime;

        let latest_txn_time = transactions
            .iter()
            .map(SignedTransaction::creation_time)
            .max()
            .expect("INTERNAL BUG: Genesis block is empty");
        let now = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap();
        now
            // We have invariant that "transaction creation time" < "block creation time"
            // See `BlockPayloadCandidate::validate_header`
            .max(latest_txn_time + Duration::from_millis(1))
            .as_millis()
            .try_into()
            .expect("INTERNAL BUG: Unix timestamp exceedes u64::MAX")
    }
}

impl BlockSignature {
    /// Peer topology index
    pub fn index(&self) -> u64 {
        self.0
    }

    /// Signature itself
    pub fn payload(&self) -> &SignatureOf<BlockHeader> {
        &self.1
    }
}

impl Display for SignedBlock {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        let SignedBlock::V1(block) = self;
        block.fmt(f)
    }
}

#[cfg(feature = "http")]
pub mod stream {
    //! Blocks for streaming API.

    use derive_more::Constructor;
    use iroha_schema::IntoSchema;
    use parity_scale_codec::{Decode, Encode};

    pub use self::model::*;
    use super::*;

    #[model]
    mod model {
        use core::num::NonZeroU64;

        use super::*;

        /// Request sent to subscribe to blocks stream starting from the given height.
        #[derive(
            Debug, Clone, Copy, Constructor, Decode, Encode, Deserialize, Serialize, IntoSchema,
        )]
        #[repr(transparent)]
        pub struct BlockSubscriptionRequest(pub NonZeroU64);

        /// Message sent by the stream producer containing block.
        #[derive(Debug, Clone, Decode, Encode, Deserialize, Serialize, IntoSchema)]
        #[repr(transparent)]
        pub struct BlockMessage(pub SignedBlock);
    }

    impl From<BlockMessage> for SignedBlock {
        fn from(source: BlockMessage) -> Self {
            source.0
        }
    }

    /// Exports common structs and enums from this module.
    pub mod prelude {
        pub use super::{BlockMessage, BlockSubscriptionRequest};
    }
}

pub mod error {
    //! Module containing errors that can occur during instruction evaluation

    pub use self::model::*;
    use super::*;

    #[model]
    mod model {
        use super::*;

        /// The reason for rejecting a transaction with new blocks.
        #[derive(
            Debug,
            Display,
            Clone,
            Copy,
            PartialEq,
            Eq,
            PartialOrd,
            Ord,
            iroha_macro::FromVariant,
            Decode,
            Encode,
            Deserialize,
            Serialize,
            IntoSchema,
        )]
        #[display(fmt = "Block was rejected during consensus")]
        #[serde(untagged)] // Unaffected by #3330 as it's a unit variant
        #[repr(transparent)]
        #[ffi_type]
        pub enum BlockRejectionReason {
            /// Block was rejected during consensus.
            ConsensusBlockRejection,
        }
    }

    #[cfg(feature = "std")]
    impl std::error::Error for BlockRejectionReason {}
}
