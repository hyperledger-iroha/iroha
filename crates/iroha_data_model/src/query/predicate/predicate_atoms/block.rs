//! This module contains predicates for block-related objects, mirroring [`crate::block`].

#[cfg(not(feature = "std"))]
use alloc::{boxed::Box, format, string::String, vec::Vec};

use iroha_crypto::HashOf;
use iroha_schema::IntoSchema;
use parity_scale_codec::{Decode, Encode};
use serde::{Deserialize, Serialize};

use super::impl_predicate_box;
use crate::{
    block::{BlockHeader, SignedBlock},
    prelude::{AccountIdPredicateBox, TransactionRejectionReason},
    query::{
        predicate::{
            predicate_ast_extensions::AstPredicateExt as _,
            predicate_combinators::{AndAstPredicate, NotAstPredicate, OrAstPredicate},
            projectors::BaseProjector,
            AstPredicate, CompoundPredicate, EvaluatePredicate, HasPredicateBox, HasPrototype,
        },
        TransactionQueryOutput,
    },
    transaction::{CommittedTransaction, SignedTransaction},
};

/// A predicate that can be applied to a [`HashOf<BlockHeader>`]
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum BlockHashPredicateBox {
    // object-specific predicates
    /// Checks if the input is equal to the expected value.
    Equals(HashOf<BlockHeader>),
}

impl_predicate_box!(HashOf<BlockHeader>: BlockHashPredicateBox);

impl EvaluatePredicate<HashOf<BlockHeader>> for BlockHashPredicateBox {
    fn applies(&self, input: &HashOf<BlockHeader>) -> bool {
        match self {
            BlockHashPredicateBox::Equals(hash) => input == hash,
        }
    }
}

/// A predicate that can be applied to a [`BlockHeader`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum BlockHeaderPredicateBox {
    // projections
    /// Checks if a predicate applies to the hash of the block header.
    Hash(BlockHashPredicateBox),
}

impl_predicate_box!(BlockHeader: BlockHeaderPredicateBox);

impl EvaluatePredicate<BlockHeader> for BlockHeaderPredicateBox {
    fn applies(&self, input: &BlockHeader) -> bool {
        match self {
            BlockHeaderPredicateBox::Hash(hash) => hash.applies(&input.hash()),
        }
    }
}

/// A predicate that can be applied to a [`SignedBlock`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum SignedBlockPredicateBox {
    // projections
    /// Checks if a predicate applies to the header of the block.
    Header(BlockHeaderPredicateBox),
}

impl_predicate_box!(SignedBlock: SignedBlockPredicateBox);

impl EvaluatePredicate<SignedBlock> for SignedBlockPredicateBox {
    fn applies(&self, input: &SignedBlock) -> bool {
        match self {
            SignedBlockPredicateBox::Header(header) => header.applies(input.header()),
        }
    }
}

/// A predicate that can be applied to a [`HashOf<SignedTransaction>`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum TransactionHashPredicateBox {
    // object-specific predicates
    /// Checks if the input is equal to the expected value.
    Equals(HashOf<SignedTransaction>),
}

impl_predicate_box!(HashOf<SignedTransaction>: TransactionHashPredicateBox);

impl EvaluatePredicate<HashOf<SignedTransaction>> for TransactionHashPredicateBox {
    fn applies(&self, input: &HashOf<SignedTransaction>) -> bool {
        match self {
            TransactionHashPredicateBox::Equals(hash) => input == hash,
        }
    }
}

/// A predicate that can be applied to a [`SignedTransaction`]
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum SignedTransactionPredicateBox {
    // projections
    /// Checks if a predicate applies to the hash of the signed transaction.
    Hash(TransactionHashPredicateBox),
    /// Checks if a predicate applies to the authority of the signed transaction.
    Authority(AccountIdPredicateBox),
}

impl_predicate_box!(SignedTransaction: SignedTransactionPredicateBox);

impl EvaluatePredicate<SignedTransaction> for SignedTransactionPredicateBox {
    fn applies(&self, input: &SignedTransaction) -> bool {
        match self {
            SignedTransactionPredicateBox::Hash(hash) => hash.applies(&input.hash()),
            SignedTransactionPredicateBox::Authority(authority) => {
                authority.applies(input.authority())
            }
        }
    }
}

// TODO: maybe we would want to have a generic `Option` predicate box & predicate
/// A predicate that can be applied to an [`Option<TransactionRejectionReason>`]
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum TransactionErrorPredicateBox {
    // object-specific predicates
    /// Checks if there was an error while applying the transaction.
    IsSome,
}

impl_predicate_box!(Option<Box<TransactionRejectionReason>>: TransactionErrorPredicateBox);

impl EvaluatePredicate<Option<Box<TransactionRejectionReason>>> for TransactionErrorPredicateBox {
    fn applies(&self, input: &Option<Box<TransactionRejectionReason>>) -> bool {
        match self {
            TransactionErrorPredicateBox::IsSome => input.is_some(),
        }
    }
}

/// A predicate that can be applied to a [`CommittedTransaction`]
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum CommittedTransactionPredicateBox {
    // projections
    /// Checks if a predicate applies to the signed transaction inside.
    Value(SignedTransactionPredicateBox),
    /// Checks if a predicate applies to the error of the transaction.
    Error(TransactionErrorPredicateBox),
}

impl_predicate_box!(CommittedTransaction: CommittedTransactionPredicateBox);

impl EvaluatePredicate<CommittedTransaction> for CommittedTransactionPredicateBox {
    fn applies(&self, input: &CommittedTransaction) -> bool {
        match self {
            CommittedTransactionPredicateBox::Value(signed_transaction) => {
                signed_transaction.applies(&input.value)
            }
            CommittedTransactionPredicateBox::Error(error) => error.applies(&input.error),
        }
    }
}

/// A predicate that can be applied to a [`TransactionQueryOutput`].
#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
pub enum TransactionQueryOutputPredicateBox {
    // projections
    /// Checks if a predicate applies to the committed transaction inside.
    Transaction(CommittedTransactionPredicateBox),
    /// Checks if a predicate applies to the hash of the block the transaction was included in.
    BlockHash(BlockHashPredicateBox),
}

impl_predicate_box!(TransactionQueryOutput: TransactionQueryOutputPredicateBox);

impl EvaluatePredicate<TransactionQueryOutput> for TransactionQueryOutputPredicateBox {
    fn applies(&self, input: &TransactionQueryOutput) -> bool {
        match self {
            TransactionQueryOutputPredicateBox::Transaction(committed_transaction) => {
                committed_transaction.applies(&input.transaction)
            }
            TransactionQueryOutputPredicateBox::BlockHash(block_hash) => {
                block_hash.applies(&input.block_hash)
            }
        }
    }
}

pub mod prelude {
    //! Re-export all predicate boxes for a glob import `(::*)`
    pub use super::{
        BlockHashPredicateBox, BlockHeaderPredicateBox, CommittedTransactionPredicateBox,
        SignedBlockPredicateBox, SignedTransactionPredicateBox, TransactionErrorPredicateBox,
        TransactionHashPredicateBox, TransactionQueryOutputPredicateBox,
    };
}

#[cfg(test)]
mod test {
    use iroha_crypto::{Hash, HashOf};

    use crate::{
        account::AccountId,
        prelude::{
            AccountIdPredicateBox, BlockHeaderPredicateBox, CompoundPredicate,
            SignedBlockPredicateBox, TransactionQueryOutputPredicateBox,
        },
        query::predicate::predicate_atoms::block::{
            BlockHashPredicateBox, CommittedTransactionPredicateBox, SignedTransactionPredicateBox,
            TransactionErrorPredicateBox, TransactionHashPredicateBox,
        },
    };

    #[test]
    fn transaction_smoke() {
        let hash = Hash::prehashed([0; 32]);
        let account_id: AccountId =
            "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland"
                .parse()
                .unwrap();

        let predicate = TransactionQueryOutputPredicateBox::build(|tx| {
            tx.block_hash.eq(HashOf::from_untyped_unchecked(hash))
                & tx.transaction.error.is_some()
                & tx.transaction.value.authority.eq(account_id.clone())
                & tx.transaction
                    .value
                    .hash
                    .eq(HashOf::from_untyped_unchecked(hash))
        });

        assert_eq!(
            predicate,
            CompoundPredicate::And(vec![
                CompoundPredicate::Atom(TransactionQueryOutputPredicateBox::BlockHash(
                    BlockHashPredicateBox::Equals(HashOf::from_untyped_unchecked(hash))
                )),
                CompoundPredicate::Atom(TransactionQueryOutputPredicateBox::Transaction(
                    CommittedTransactionPredicateBox::Error(TransactionErrorPredicateBox::IsSome)
                )),
                CompoundPredicate::Atom(TransactionQueryOutputPredicateBox::Transaction(
                    CommittedTransactionPredicateBox::Value(
                        SignedTransactionPredicateBox::Authority(AccountIdPredicateBox::Equals(
                            account_id.clone()
                        ))
                    )
                )),
                CompoundPredicate::Atom(TransactionQueryOutputPredicateBox::Transaction(
                    CommittedTransactionPredicateBox::Value(SignedTransactionPredicateBox::Hash(
                        TransactionHashPredicateBox::Equals(HashOf::from_untyped_unchecked(hash))
                    ))
                )),
            ])
        );
    }

    #[test]
    fn block_smoke() {
        let hash = Hash::prehashed([0; 32]);

        let predicate = SignedBlockPredicateBox::build(|block| {
            block.header.hash.eq(HashOf::from_untyped_unchecked(hash))
        });

        assert_eq!(
            predicate,
            CompoundPredicate::Atom(SignedBlockPredicateBox::Header(
                BlockHeaderPredicateBox::Hash(BlockHashPredicateBox::Equals(
                    HashOf::from_untyped_unchecked(hash)
                ))
            ))
        );
    }
}
