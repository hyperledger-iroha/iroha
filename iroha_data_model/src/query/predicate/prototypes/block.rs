//! Account-related prototypes, mirroring types in [`crate::block`].

use core::marker::PhantomData;

use iroha_crypto::HashOf;

use super::impl_prototype;
use crate::{
    block::BlockHeader,
    query::predicate::{
        predicate_ast_extensions::AstPredicateExt,
        predicate_atoms::block::{
            BlockHashPredicateBox, BlockHeaderPredicateBox, CommittedTransactionPredicateBox,
            SignedBlockPredicateBox, SignedTransactionPredicateBox, TransactionErrorPredicateBox,
            TransactionHashPredicateBox, TransactionQueryOutputPredicateBox,
        },
        predicate_combinators::NotAstPredicate,
        projectors::{
            BlockHeaderHashProjector, CommittedTransactionErrorProjector,
            CommittedTransactionValueProjector, ObjectProjector, SignedBlockHeaderProjector,
            SignedTransactionAuthorityProjector, SignedTransactionHashProjector,
            TransactionQueryOutputBlockHashProjector, TransactionQueryOutputTransactionProjector,
        },
        prototypes::account::AccountIdPrototype,
        AstPredicate, HasPrototype,
    },
    transaction::SignedTransaction,
};

/// A prototype of [`HashOf<BlockHeader>`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct BlockHashPrototype<Projector> {
    phantom: PhantomData<Projector>,
}

impl_prototype!(BlockHashPrototype: BlockHashPredicateBox);

impl<Projector> BlockHashPrototype<Projector>
where
    Projector: ObjectProjector<Input = BlockHashPredicateBox>,
{
    /// Creates a predicate that checks if the hash equals the expected value.
    pub fn eq(
        &self,
        expected: HashOf<BlockHeader>,
    ) -> Projector::ProjectedPredicate<BlockHashPredicateBox> {
        Projector::project_predicate(BlockHashPredicateBox::Equals(expected))
    }
}

/// A prototype of [`BlockHeader`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct BlockHeaderPrototype<Projector> {
    /// Build a predicate on hash of this [`BlockHeader`]
    pub hash: BlockHashPrototype<BlockHeaderHashProjector<Projector>>,
}

impl_prototype!(BlockHeaderPrototype: BlockHeaderPredicateBox);

/// A prototype of [`crate::block::SignedBlock`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct SignedBlockPrototype<Projector> {
    /// Build a predicate on header of this [`crate::block::SignedBlock`]
    pub header: BlockHeaderPrototype<SignedBlockHeaderProjector<Projector>>,
}

impl_prototype!(SignedBlockPrototype: SignedBlockPredicateBox);

/// A prototype of [`HashOf<SignedTransaction>`]
#[derive(Default, Copy, Clone)]
pub struct TransactionHashPrototype<Projector> {
    phantom: PhantomData<Projector>,
}

impl_prototype!(TransactionHashPrototype: TransactionHashPredicateBox);

impl<Projector> TransactionHashPrototype<Projector>
where
    Projector: ObjectProjector<Input = TransactionHashPredicateBox>,
{
    /// Creates a predicate that checks if the hash equals the expected value.
    pub fn eq(
        &self,
        expected: HashOf<SignedTransaction>,
    ) -> Projector::ProjectedPredicate<TransactionHashPredicateBox> {
        Projector::project_predicate(TransactionHashPredicateBox::Equals(expected))
    }
}

/// A prototype of [`SignedTransaction`]
#[derive(Default, Copy, Clone)]
pub struct SignedTransactionPrototype<Projector> {
    /// Build a predicate on hash of this [`SignedTransaction`]
    pub hash: TransactionHashPrototype<SignedTransactionHashProjector<Projector>>,
    /// Build a predicate on the transaction authority
    pub authority: AccountIdPrototype<SignedTransactionAuthorityProjector<Projector>>,
}

impl_prototype!(SignedTransactionPrototype: SignedTransactionPredicateBox);

/// A prototype of [`Option<crate::transaction::error::TransactionRejectionReason>`]
#[derive(Default, Copy, Clone)]
pub struct TransactionErrorPrototype<Projector> {
    phantom: PhantomData<Projector>,
}

impl_prototype!(TransactionErrorPrototype: TransactionErrorPredicateBox);

impl<Projector> TransactionErrorPrototype<Projector>
where
    Projector: ObjectProjector<Input = TransactionErrorPredicateBox>,
{
    /// Creates a predicate that checks if there is an error.
    pub fn is_some(&self) -> Projector::ProjectedPredicate<TransactionErrorPredicateBox> {
        Projector::project_predicate(TransactionErrorPredicateBox::IsSome)
    }

    /// Creates a predicate that checks if there is no error.
    pub fn is_none(
        &self,
    ) -> NotAstPredicate<Projector::ProjectedPredicate<TransactionErrorPredicateBox>> {
        Projector::project_predicate(TransactionErrorPredicateBox::IsSome).not()
    }
}

/// A prototype of [`crate::transaction::CommittedTransaction`]
#[derive(Default, Copy, Clone)]
pub struct CommittedTransactionPrototype<Projector> {
    /// Build a predicate on the signed transaction inside
    pub value: SignedTransactionPrototype<CommittedTransactionValueProjector<Projector>>,
    /// Build a predicate on the transaction error
    pub error: TransactionErrorPrototype<CommittedTransactionErrorProjector<Projector>>,
}

impl_prototype!(CommittedTransactionPrototype: CommittedTransactionPredicateBox);

/// A prototype of [`crate::query::TransactionQueryOutput`] for predicate construction.
#[derive(Default, Copy, Clone)]
pub struct TransactionQueryOutputPrototype<Projector> {
    /// Build a predicate on the transaction inside
    pub transaction:
        CommittedTransactionPrototype<TransactionQueryOutputTransactionProjector<Projector>>,
    /// Build a predicate on the block hash inside
    pub block_hash: BlockHashPrototype<TransactionQueryOutputBlockHashProjector<Projector>>,
}

impl_prototype!(TransactionQueryOutputPrototype: TransactionQueryOutputPredicateBox);
