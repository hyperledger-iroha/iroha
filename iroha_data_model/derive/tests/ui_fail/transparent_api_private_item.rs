use iroha_data_model::block::BlockPayload;

fn main() {}
