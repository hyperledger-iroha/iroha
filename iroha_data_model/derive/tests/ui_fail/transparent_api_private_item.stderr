error[E0603]: struct `BlockPayload` is private
 --> tests/ui_fail/transparent_api_private_item.rs:1:30
  |
1 | use iroha_data_model::block::BlockPayload;
  |                              ^^^^^^^^^^^^ private struct
  |
note: the struct `BlockPayload` is defined here
 --> $WORKSPACE/data_model/src/block.rs
  |
  | pub use self::model::*;
  |         ^^^^^^^^^^^
